node/core/dispute-coordinator/src/real/ordering/mod.rs

```diff
@@ -311,12 +334,39 @@ where
 			rx.await??
 		};
 
+		// Try to fetch block numbers for the obtained block hashes.
+		let mut receivers = Vec::with_capacity(hashes.len());
 		for hash in &hashes {
-			if *hash == last_observed_hash {
+			let (tx, rx) = oneshot::channel();
+			sender.send_message(ChainApiMessage::BlockNumber(*hash, tx).into()).await;
```

Comment:
We can actually avoid calling `BlockNumber` for each ancestor, as the parent of a block will always have child.block_number - 1. Thus if we know the block number of the starting leaf, we don't need to lookup any block numbers.